File: src/main/java/com/marcpg/pillarperil/event/PlayerEvents.java
package com.marcpg.pillarperil.event;

import com.marcpg.pillarperil.PillarPlayer;
import com.marcpg.pillarperil.game.Game;
import com.marcpg.pillarperil.game.util.GameManager;
import com.marcpg.pillarperil.generation.Generator;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.PlayerDeathEvent;
import org.bukkit.event.player.PlayerMoveEvent;
import org.jetbrains.annotations.NotNull;

public class PlayerEvents implements Listener {
    @EventHandler(ignoreCancelled = true)
    public void onPlayerDeath(@NotNull PlayerDeathEvent event) {
        PillarPlayer player = GameManager.player(event.getPlayer(), true);
        if (player != null) {
            if (player.player.getKiller() != null) {
                PillarPlayer killer = player.game.player(player.player.getKiller(), false);
                if (killer != null) killer.addKill();
            }

            player.game.eliminate(player);
        }
    }

    @EventHandler(ignoreCancelled = true)
    public void onPlayerMove(@NotNull PlayerMoveEvent event) {
        if (event.getTo().y() < Generator.deathHeight) {
            Game game = GameManager.game(event.getPlayer(), true);
            if (game != null)
                event.getPlayer().setHealth(0.0);
        }
    }
}

File: src/main/kotlin/com/marcpg/pillarperil/event/PlayerEvents.kt
package com.marcpg.pillarperil.event

import com.marcpg.libpg.util.bukkitRunLater
import com.marcpg.pillarperil.game.util.GameManager
import com.marcpg.pillarperil.game.util.QueueManager
import com.marcpg.pillarperil.util.Configuration
import com.marcpg.pillarperil.util.QueueMethod
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.entity.PlayerDeathEvent
import org.bukkit.event.player.PlayerJoinEvent
import org.bukkit.event.player.PlayerMoveEvent
import org.bukkit.event.player.PlayerQuitEvent

object PlayerEvents : Listener {
    @EventHandler(ignoreCancelled = true)
    fun onPlayerDeath(event: PlayerDeathEvent) {
        val player = GameManager.player(event.player) ?: return

        if (event.player.killer != null)
            player.game.player(event.player.killer!!, false)?.kills++

        player.game.eliminate(player)
    }

    @EventHandler(ignoreCancelled = true)
    fun onPlayerMove(event: PlayerMoveEvent) {
        if (event.to.y < Configuration.deathHeight) {
            GameManager.player(event.player) ?: return
            event.player.health = 0.0
        }
    }

    @EventHandler(ignoreCancelled = true)
    fun onPlayerJoin(event: PlayerJoinEvent) {
        if (Configuration.queueMethod == QueueMethod.AUTO)
            bukkitRunLater(20L) { if (event.player.isOnline) QueueManager.add(event.player) } // Wait 1 second before rejoining queue.
    }

    @EventHandler(ignoreCancelled = true)
    fun onPlayerQuit(event: PlayerQuitEvent) {
        QueueManager.remove(event.player)
    }
}
